### ISSUE TITLE
MSVC 2015 Compiler crash when building HPX

### ISSUE THREAD

**Issue by Naios** [STEllAR-GROUP/hpx#2574]
On revision 5609f1cddb335fb1f MSVC 2015 with Update 3 (v14.0.25431.01) crashes when building HPX:

```
hpx\src\runtime\agas\big_boot_barrier.cpp(289): fatal error C1001: An internal error has occurred in the compiler.
```

Probably this is related to the `operator&` in file [`src/runtime/agas/big_boot_barrier.cpp`](https://github.com/STEllAR-GROUP/hpx/blame/master/src/runtime/agas/big_boot_barrier.cpp) and the following lines:

- `ar & typenames`
- `ar & num_localities`
- `ar & used_cores`

```cpp
    template <typename Archive>
    void serialize(Archive & ar, const unsigned int)
    {
        ar & endpoints;
        ar & primary_ns_ptr;
        ar & symbol_ns_ptr;
        ar & cores_needed;
        ar & num_threads;
        ar & hostname;
        ar & typenames;
     // ^^^^^^^^^^^^^^^
        ar & prefix;
    }
```
... and ...

```cpp
    template <typename Archive>
    void serialize(Archive & ar, const unsigned int)
    {
        ar & prefix;
        ar & agas_locality;
        ar & locality_ns_address;
        ar & primary_ns_address;
        ar & component_ns_address;
        ar & symbol_ns_address;
        ar & num_localities;
     // ^^^^^^^^^^^^^^^
        ar & used_cores;
     // ^^^^^^^^^^^^^^^
        ar & agas_endpoints;
        ar & ids;
        ar & endpoints;
    }
```

**Comment by hkaiser** [comment]
Can you give use more details on what cmake opetion you have used, please? Also, what build type is failing (Debug, Release, etc.)?

FWIW, I'm using this version of VS to develop HPX and I have not seen this compiler error before.

**Comment by Naios** [comment]
It's the *Release* build type which is failing, the *Debug* build succeeds.

CMake was run with the `VisualStudio 14 Win64` generator using **boost 1.63 x64** and `-DHPX_WITH_HWLOC=OFF`.
My CMake cache is the following: [CMakeCache.txt](https://github.com/STEllAR-GROUP/hpx/files/886504/CMakeCache.txt).

**Comment by hkaiser** [comment]
Thanks, I have it reproduced and will investigate. As a workaround, please build with `RelWithDebInfo`. We know that works.

**Comment by hkaiser** [comment]
@Naios Could you please verify whether #2585 solves your issue and report back?

**Comment by Naios** [comment]
@hkaiser I can confirm that the fix provided by the `fixing_2574` (#2585) branch solves this issue.
